File: transportable-udfs-spark/src/main/scala/com/linkedin/transport/spark/StdUdfWrapper.scala
/**
 * Copyright 2019 LinkedIn Corporation. All rights reserved.
 * Licensed under the BSD-2 Clause license.
 * See LICENSE in the project root for license information.
 */
package com.linkedin.transport.spark

import java.io.{IOException, ObjectStreamException}
import java.nio.file.Paths
import java.util.List

import com.linkedin.transport.api.StdFactory
import com.linkedin.transport.api.data.{PlatformData, StdData}
import com.linkedin.transport.api.udf._
import com.linkedin.transport.spark.typesystem.SparkTypeInference
import com.linkedin.transport.utils.FileSystemUtils
import org.apache.spark.SparkFiles
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.catalyst.expressions.Expression
import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
import org.apache.spark.sql.types.DataType

abstract class StdUdfWrapper(_expressions: Seq[Expression]) extends Expression
  with CodegenFallback with Serializable {

  @transient private var _stdFactory: StdFactory = _
  @transient private var _stdUdf: StdUDF = _
  @transient private var _requiredFilesProcessed: Boolean = false
  @transient private var _outputDataType: DataType = _
  private var _nullableArguments: Array[Boolean] = _
  private var _distributedCacheFiles: Array[String] = _

  override def nullable: Boolean = true

  override def dataType: DataType = {
    if (_outputDataType != null) _outputDataType else initialize()
  }

  private def initialize(): DataType = {
    val sparkTypeInference = new SparkTypeInference
    sparkTypeInference.compile(children.map(_.dataType).toArray, getStdUdfImplementations, getTopLevelUdfClass)
    _stdFactory = sparkTypeInference.getStdFactory
    _stdUdf = sparkTypeInference.getStdUdf
    _nullableArguments = _stdUdf.getAndCheckNullableArguments
    _stdUdf.init(_stdFactory)
    getRequiredFiles()
    _requiredFilesProcessed = false
    _outputDataType = sparkTypeInference.getOutputDataType
    _outputDataType
  }

  override def children: Seq[Expression] = _expressions


  // Suppressing magic number warming since the number match is required to cast it into the coresponding StdUDF
  // scalastyle:off magic.number
  private def getRequiredFiles(): Unit = { // scalastyle:ignore cyclomatic.complexity
    if (_distributedCacheFiles == null) {
      val wrappedConstants = checkNullsAndWrapConstants()
      // If wrappedConstants is null, it means there were non-nullable constants whose value was evaluated to be null
      // Hence we do not call user's getRequiredFiles(). Also in such a case, null checks in eval will also fail and
      // user's eval will never be called, so there is no need to getRequiredFiles() anyway.
      if (wrappedConstants != null) {
        val requiredFiles = wrappedConstants.length match {
          case 0 =>
            _stdUdf.asInstanceOf[StdUDF0[StdData]].getRequiredFiles()
          case 1 =>
            _stdUdf.asInstanceOf[StdUDF1[StdData, StdData]].getRequiredFiles(wrappedConstants(0))
          case 2 =>
            _stdUdf.asInstanceOf[StdUDF2[StdData, StdData, StdData]].getRequiredFiles(wrappedConstants(0),
              wrappedConstants(1))
          case 3 =>
            _stdUdf.asInstanceOf[StdUDF3[StdData, StdData, StdData, StdData]].getRequiredFiles(wrappedConstants(0),
              wrappedConstants(1), wrappedConstants(2))
          case 4 =>
            _stdUdf.asInstanceOf[StdUDF4[StdData, StdData, StdData, StdData, StdData]].getRequiredFiles(wrappedConstants(0),
              wrappedConstants(1), wrappedConstants(2), wrappedConstants(3))
          case 5 =>
            _stdUdf.asInstanceOf[StdUDF5[StdData, StdData, StdData, StdData, StdData, StdData]].getRequiredFiles(wrappedConstants(0),
              wrappedConstants(1), wrappedConstants(2), wrappedConstants(3), wrappedConstants(4))
          case 6 =>
            _stdUdf.asInstanceOf[StdUDF6[StdData, StdData, StdData, StdData, StdData, StdData, StdData]].getRequiredFiles(wrappedConstants(0),
              wrappedConstants(1), wrappedConstants(2), wrappedConstants(3), wrappedConstants(4), wrappedConstants(5))
          case 7 =>
            _stdUdf.asInstanceOf[StdUDF7[StdData, StdData, StdData, StdData, StdData, StdData, StdData, StdData]].getRequiredFiles(wrappedConstants(0),
              wrappedConstants(1), wrappedConstants(2), wrappedConstants(3), wrappedConstants(4), wrappedConstants(5), wrappedConstants(6))
          case 8 =>
            _stdUdf.asInstanceOf[StdUDF8[StdData, StdData, StdData, StdData, StdData, StdData, StdData, StdData, StdData]].getRequiredFiles(wrappedConstants(0),
              wrappedConstants(1), wrappedConstants(2), wrappedConstants(3), wrappedConstants(4), wrappedConstants(5), wrappedConstants(6), wrappedConstants(7))
          case _ =>
            throw new UnsupportedOperationException("getRequiredFiles not yet supported for StdUDF" + _expressions.length)
        }

        lazy val sparkContext = SparkSession.builder().getOrCreate().sparkContext
        _distributedCacheFiles = requiredFiles.map(file => {
          try {
            val resolvedFile = FileSystemUtils.resolveLatest(file)
            // TODO: Currently does not support adding of files with same file name. E.g dirA/file.txt dirB/file.txt
            sparkContext.addFile(resolvedFile)
            resolvedFile
          } catch {
            case e: IOException =>
              throw new RuntimeException("Failed to resolve path: [" + file + "].", e)
          }
        })
      }
    }
  } // scalastyle:on magic.number

  private final def checkNullsAndWrapConstants(): Array[StdData] = {
    val wrappedConstants = new Array[StdData](_expressions.length)
    for (i <- _expressions.indices) {
      val constantValue = if (_expressions(i).foldable) _expressions(i).eval() else null
      if (!_nullableArguments(i) && _expressions(i).foldable && constantValue == null) {
        // constant is defined as non nullable and value is null, so return early
        return null // scalastyle:ignore return
      }
      wrappedConstants(i) = SparkWrapper.createStdData(constantValue, _expressions(i).dataType)
    }
    wrappedConstants
  }

  // Suppressing magic number warming since the number match is required to cast it into the corresponding StdUDF
  // scalastyle:off magic.number
  override def eval(input: InternalRow): Any = { // scalastyle:ignore cyclomatic.complexity
    val wrappedArguments = checkNullsAndWrapArguments(input)
    // If wrappedArguments is null, it means there were non-nullable arguments whose value was evaluated to be null
    // So we do not call user's eval()
    if (wrappedArguments == null) {
      null
    } else {
      if (!_requiredFilesProcessed) {
        processRequiredFiles()
      }
      val stdResult = wrappedArguments.length match {
        case 0 =>
          _stdUdf.asInstanceOf[StdUDF0[StdData]].eval()
        case 1 =>
          _stdUdf.asInstanceOf[StdUDF1[StdData, StdData]].eval(wrappedArguments(0))
        case 2 =>
          _stdUdf.asInstanceOf[StdUDF2[StdData, StdData, StdData]].eval(wrappedArguments(0), wrappedArguments(1))
        case 3 =>
          _stdUdf.asInstanceOf[StdUDF3[StdData, StdData, StdData, StdData]].eval(wrappedArguments(0), wrappedArguments(1),
            wrappedArguments(2))
        case 4 =>
          _stdUdf.asInstanceOf[StdUDF4[StdData, StdData, StdData, StdData, StdData]].eval(wrappedArguments(0),
            wrappedArguments(1), wrappedArguments(2), wrappedArguments(3))
        case 5 =>
          _stdUdf.asInstanceOf[StdUDF5[StdData, StdData, StdData, StdData, StdData, StdData]].eval(wrappedArguments(0),
            wrappedArguments(1), wrappedArguments(2), wrappedArguments(3), wrappedArguments(4))
        case 6 =>
          _stdUdf.asInstanceOf[StdUDF6[StdData, StdData, StdData, StdData, StdData, StdData, StdData]].eval(wrappedArguments(0),
            wrappedArguments(1), wrappedArguments(2), wrappedArguments(3), wrappedArguments(4), wrappedArguments(5))
        case 7 =>
          _stdUdf.asInstanceOf[StdUDF7[StdData, StdData, StdData, StdData, StdData, StdData, StdData, StdData]].eval(wrappedArguments(0),
            wrappedArguments(1), wrappedArguments(2), wrappedArguments(3), wrappedArguments(4), wrappedArguments(5),
            wrappedArguments(6))
        case 8 =>
          _stdUdf.asInstanceOf[StdUDF8[StdData, StdData, StdData, StdData, StdData, StdData, StdData, StdData, StdData]].eval(wrappedArguments(0),
            wrappedArguments(1), wrappedArguments(2), wrappedArguments(3), wrappedArguments(4), wrappedArguments(5),
            wrappedArguments(6), wrappedArguments(7))
        case _ =>
          throw new UnsupportedOperationException("eval not yet supported for StdUDF" + _expressions.length)
      }

      if (stdResult == null) null else stdResult.asInstanceOf[PlatformData].getUnderlyingData
    }
  } // scalastyle:on magic.number


  private final def checkNullsAndWrapArguments(input: InternalRow): Array[StdData] = {
    val wrappedArguments = new Array[StdData](_expressions.length)
    for (i <- _expressions.indices) {
      val evaluatedExpression = _expressions(i).eval(input)
      if(!_nullableArguments(i) && evaluatedExpression == null) {
        // argument is defined as non nullable and value is null, so return early
        return null // scalastyle:ignore return
      }
      wrappedArguments(i) = SparkWrapper.createStdData(evaluatedExpression, _expressions(i).dataType)
    }
    wrappedArguments
  }

  private final def processRequiredFiles(): Unit = {
    if (!_requiredFilesProcessed) {
      val localFiles = _distributedCacheFiles.map(file => {
        try {
          SparkFiles.get(Paths.get(file).getFileName.toString)
        } catch {
          case e: IOException =>
            throw new RuntimeException("Failed to resolve path: [" + file + "].", e)
        }
      })
      _stdUdf.processRequiredFiles(localFiles)
      _requiredFilesProcessed = true
    }
  }

  protected def getStdUdfImplementations: List[_ <: StdUDF]

  protected def getTopLevelUdfClass: Class[_ <: TopLevelStdUDF]

  override def makeCopy(newArgs: Array[AnyRef]): Expression = {
    val newInstance = super.makeCopy(newArgs).asInstanceOf[StdUdfWrapper]
    if (newInstance != null) {
      newInstance._stdFactory = _stdFactory
      newInstance._stdUdf = _stdUdf
      newInstance._requiredFilesProcessed = _requiredFilesProcessed
      newInstance._outputDataType = _outputDataType
      newInstance._nullableArguments = _nullableArguments
      newInstance._distributedCacheFiles = _distributedCacheFiles
    }
    newInstance
  }

  @throws(classOf[ObjectStreamException])
  private def readResolve(): Object = {
    initialize()
    this
  }
}

File: transportable-udfs-avro/src/main/java/com/linkedin/transport/avro/StdUdfWrapper.java
/**
 * Copyright 2019 LinkedIn Corporation. All rights reserved.
 * Licensed under the BSD-2 Clause license.
 * See LICENSE in the project root for license information.
 */
package com.linkedin.transport.avro;

import com.linkedin.transport.api.StdFactory;
import com.linkedin.transport.api.data.PlatformData;
import com.linkedin.transport.api.udf.StdUDF;
import com.linkedin.transport.api.udf.StdUDF0;
import com.linkedin.transport.api.udf.StdUDF1;
import com.linkedin.transport.api.udf.StdUDF2;
import com.linkedin.transport.api.udf.StdUDF3;
import com.linkedin.transport.api.udf.StdUDF4;
import com.linkedin.transport.api.udf.StdUDF5;
import com.linkedin.transport.api.udf.StdUDF6;
import com.linkedin.transport.api.udf.StdUDF7;
import com.linkedin.transport.api.udf.StdUDF8;
import com.linkedin.transport.api.udf.TopLevelStdUDF;
import com.linkedin.transport.avro.typesystem.AvroTypeInference;
import java.util.List;
import java.util.stream.IntStream;
import org.apache.avro.Schema;


/**
 * Base class for all Avro Standard UDFs. It provides a standard way of type validation, binding, and output type
 * inference through its initialize() method.
 */
public abstract class StdUdfWrapper {

  protected Schema[] _inputSchemas;
  protected StdUDF _stdUdf;
  protected boolean _requiredFilesProcessed;
  protected StdFactory _stdFactory;
  private boolean[] _nullableArguments;
  private Object[] _args;

  /**
   * Given input schemas, this method matches them to the expected type signatures, and finds bindings to the
   * generic parameters. Once the generic parameter bindings are known, the method infers the output type (in the form
   * of an Avro schema) by substituting the binding values in the output type signature.
   * signature and
   * @param arguments Input Avro Schemas of UDF parameters.
   * @return Inferred output Avro Schema.
   */
  public Schema initialize(Schema[] arguments) {
    AvroTypeInference avroTypeInference = new AvroTypeInference();
    avroTypeInference.compile(arguments, getStdUdfImplementations(), getTopLevelUdfClass());
    _inputSchemas = avroTypeInference.getInputDataTypes();
    _stdFactory = avroTypeInference.getStdFactory();
    _stdUdf = avroTypeInference.getStdUdf();
    _nullableArguments = _stdUdf.getAndCheckNullableArguments();
    _stdUdf.init(_stdFactory);
    _requiredFilesProcessed = false;
    createStdData();
    return avroTypeInference.getOutputDataType();
  }

  protected boolean containsNullValuedNonNullableArgument(Object[] arguments) {
    for (int i = 0; i < arguments.length; i++) {
      if (arguments[i] == null && !_nullableArguments[i]) {
        return true;
      }
    }
    return false;
  }

  protected Object wrap(Object avroObject, Schema inputSchema, Object stdData) {
    switch (inputSchema.getType()) {
      case INT:
      case LONG:
      case BOOLEAN:
        return avroObject;
      case STRING:
        return avroObject == null? null : avroObject.toString();
      case ARRAY:
      case MAP:
      case RECORD:
        if (avroObject != null) {
          ((PlatformData) stdData).setUnderlyingData(avroObject);
          return stdData;
        } else {
          return null;
        }
      case NULL:
        return null;
      default:
        throw new RuntimeException("Unrecognized Avro Schema: " + inputSchema.getClass());
    }
  }

  protected abstract List<? extends StdUDF> getStdUdfImplementations();

  protected abstract Class<? extends TopLevelStdUDF> getTopLevelUdfClass();

  protected void createStdData() {
    _args = new Object[_inputSchemas.length];
    for (int i = 0; i < _inputSchemas.length; i++) {
      _args[i] = AvroWrapper.createStdData(null, _inputSchemas[i]);
    }
  }

  private Object[] wrapArguments(Object[] arguments) {
    return IntStream.range(0, _args.length).mapToObj(
        i -> wrap(arguments[i], _inputSchemas[i], _args[i])
    ).toArray(Object[]::new);
  }

  public Object evaluate(Object[] arguments) {
    if (containsNullValuedNonNullableArgument(arguments)) {
      return null;
    }
    Object[] args = wrapArguments(arguments);
    Object result;
    switch (args.length) {
      case 0:
        result = ((StdUDF0) _stdUdf).eval();
        break;
      case 1:
        result = ((StdUDF1) _stdUdf).eval(args[0]);
        break;
      case 2:
        result = ((StdUDF2) _stdUdf).eval(args[0], args[1]);
        break;
      case 3:
        result = ((StdUDF3) _stdUdf).eval(args[0], args[1], args[2]);
        break;
      case 4:
        result = ((StdUDF4) _stdUdf).eval(args[0], args[1], args[2], args[3]);
        break;
      case 5:
        result = ((StdUDF5) _stdUdf).eval(args[0], args[1], args[2], args[3], args[4]);
        break;
      case 6:
        result = ((StdUDF6) _stdUdf).eval(args[0], args[1], args[2], args[3], args[4], args[5]);
        break;
      case 7:
        result = ((StdUDF7) _stdUdf).eval(args[0], args[1], args[2], args[3], args[4], args[5], args[6]);
        break;
      case 8:
        result = ((StdUDF8) _stdUdf).eval(args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7]);
        break;
      default:
        throw new UnsupportedOperationException("eval not yet supported for StdUDF" + args.length);
    }
    return AvroWrapper.getPlatformData(result);
  }
}
